.gitlab/conf/host-configs/blueos_3_ppc64le_ib/nvcc_9_2_xl_2019_04_19.cmake

```diff
@@ -0,0 +1,11 @@
+###############################################################################
+# Copyright (c) 2016-19, Lawrence Livermore National Security, LLC
+# and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (BSD-3-Clause)
+###############################################################################
+
+set(CMAKE_CXX_COMPILER "/usr/tce/packages/xl/xl-2019.04.19/bin/xlc++_r" CACHE PATH "" )
+
+set(CUDA_TOOLKIT_ROOT_DIR "/usr/tce/packages/cuda/cuda-9.2.148" CACHE PATH "" )
```

Comment:
@adrienbernede Everything looks great! This is a good start.

"However, there is in RAJA a third level of configuration: at script time. This is not handled yet. And I would like it to be discussed."

I'm not entirely sure what you mean by this, but in general I think the RAJA level of modularity is good. In fact, we could attempt to make it even _more_ modular by separating the combinations of compiler paths. For example, this particular file could be chopped into an "nvcc_9_2.cmake" and an "xl_2019_04_19.cmake". This would become rather complex though, because we would also need to separate host-configs/.../nvcc_X_xl_X.cmake into 2 files as well, and the gitlab/build_and_test.sh regex would become more difficult.

Reply:
"However, there is in RAJA a third level of configuration: at script time. This is not handled yet. And I would like it to be discussed."

What I meant there is that in RAJA build scripts the configuration is determined using 3 sources of information:
- the toolchain config files (in .gitlab/conf/host-configs),
- the project config files (in host-configs),
- and the script itself, providing definitions in the CMake command line. 
This last "level of configuration" is not present in the generic `scripts/gitlab/build_and_test.sh`. This means that the CI builds are not the same as the scripted builds at the moment. This is important to notice, and I would like to discuss the ways to address this.

You are right, the current design is a trade-off between modularity and simplicity. What I envision next is to generate or retrieve the toolchain config-files from a list of desired toolchain. This way we could have a single, let’s say yaml, file were the configurations are listed in a compact and readable way.

Reply:
If I understand you correctly, you're saying that cmake options like "-DENABLE_OPENMP=On" or "-DENABLE_CUDA=Off" (and all such combinations) are difficult to encode in gitlab/build_and_test.sh? This would indeed be a problem.

One way to solve this (possibly difficult) would be to use regex rules. For instance, if we detect nvcc in the config then include "-DENABLE_CUDA=On", and depending on the nvcc version number set "-DCUDA_ARCH=sm_something". This will devolve into a build_and_test.sh that is difficult to grok though.

We could also have separate build_and_test.sh's for different general cases, for example:

- build_and_test_nvcc10.sh
- build_and_test_nvcc9.sh
- build_and_test_nvcc8.sh
- build_and_test.sh
- build_and_test_omp.sh
- build_and_test_omptarget.sh

Each of these would have static cmake lines, and read in their corresponding test cases.

Reply:
I think the examples you gave are not the ones that would cause the most troubles.
In fact, _if there is a simple rule_ that dictates with which toolchain we want OpenMP and or Cuda, and with which nvcc version to associate a sm_xx arch, then we could deal with it.

For example, the build_and_test.sh script currently expects a unique matching host-configs file. But we could imagine using all the matching ones.  

For example, both `nvcc_X_clang_X.cmake` and `nvcc_9_2_X.cmake` would match nvcc_9_2_clang_6_0_0.cmake. And `nvcc_9_2_X.cmake` would set `-DCUDA_ARCH=sm_something`.

However, this proposal cannot deal with two different configurations applying to the same target (Debug/Release/EA). I have some twisted idea about how to do that, though.